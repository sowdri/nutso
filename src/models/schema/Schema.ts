import { ArraySchema } from "./ArraySchema";
import { BooleanSchema } from "./BooleanSchema";
import { DateSchema } from "./DateSchema";
import { NumberSchema } from "./NumberSchema";
import { ObjectSchema } from "./ObjectSchema";
import { StringSchema } from "./StringSchema";
import { TupleSchema } from "./TupleSchema";

/**
 * # Terminology (Only required to refer the code)
 * T => Type of the value currently being validated
 * K => key => Type of Key
 * R => root => The root of the object that is being validated
 * P => parent => The parent of the object being validated. If for [K in keyof T],
 */

export type Schema<T, R = T, P = unknown> = T extends string
  ? StringSchema<R, P>
  : T extends number
  ? NumberSchema<R, P>
  : T extends Date
  ? DateSchema<R, P>
  : T extends boolean
  ? BooleanSchema<R, P>
  : T extends Array<infer E>
  ? ArraySchema<E, T, R, P>
  : T extends [infer U, ...unknown[]]
  ? TupleSchema<U, R, P>
  : T extends Object
  ? ObjectSchema<T, R, P>
  : never;
