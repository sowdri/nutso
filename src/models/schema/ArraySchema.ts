import { OptionalFlag } from "../OptionalFlag";
import { Schema } from "./Schema";

/**
 * E => Element
 * T => Type (this is the array)
 * P => Parent of the array
 *
 * Look at this type, we need all 3 types
 */
export type ArraySchema<E, T extends E[], R, P = unknown> = {
  type: "array";
  optional?: OptionalFlag<R, P>;
  minItems?: number;
  maxItems?: number;
  items: Schema<E, R>; // TODO check how this works in schema definition
};
