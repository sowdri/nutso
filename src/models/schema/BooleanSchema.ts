import { OptionalFlag } from "../OptionalFlag";
import { ValidationFn } from "../ValidationFn";

export type BooleanSchema<R = boolean, P = unknown> = {
  type: "boolean";
  optional?: OptionalFlag<R, P>;
  validationFn?: ValidationFn<boolean, R, P>;
};
