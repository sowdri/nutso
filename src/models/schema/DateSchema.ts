import { OptionalFlag } from "../OptionalFlag";
import { ValidationFn } from "../ValidationFn";

export type DateSchema<R = Date, P = unknown> = {
  type: "date";
  optional?: OptionalFlag<R, P>;
  validationFn?: ValidationFn<Date, R, P>;
};
