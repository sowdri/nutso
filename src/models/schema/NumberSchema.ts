import { OptionalFlag } from "../OptionalFlag";
import { ValidationFn } from "../ValidationFn";

export type NumberSchema<R = number, P = unknown> = {
  type: "number";
  min?: number;
  max?: number;
  pattern?: RegExp;
  optional?: OptionalFlag<R, P>;
  validationFn?: ValidationFn<number, R, P>;
};
