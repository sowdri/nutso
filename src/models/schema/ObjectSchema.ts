import { Schema } from "./Schema";
import { OptionalFlag } from "../OptionalFlag";

export type ObjectSchema<T, R = T, P = unknown> = {
  type: "object";
  optional?: OptionalFlag<R, P>;
  properties: {
    [K in keyof T]: Schema<T[K], R, T>;
  };
};
