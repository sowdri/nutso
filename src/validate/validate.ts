import { Result } from "../models/result/Result";
import { BooleanSchema } from "../models/schema/BooleanSchema";
import { DateSchema } from "../models/schema/DateSchema";
import { NumberSchema } from "../models/schema/NumberSchema";
import { Schema } from "../models/schema/Schema";
import { StringSchema } from "../models/schema/StringSchema";
import { validateArray } from "./validateArray";
import { validateBoolean } from "./validateBoolean";
import { validateDate } from "./validateDate";
import { validateNumber } from "./validateNumber";
import { validateObject } from "./validateObject";
import { validateString } from "./validateString";

/**
 * This is an internal API that takes the root as an argument
 * and called recursively from `validateObject` and `validateArray` methods
 *
 * @param value
 * @param parent
 * @param schema
 */
export const _validate = <T, P>(args: { value: T | null; parent: P; schema: Schema<T, P> }): Result<T> => {
  //

  if (!args.schema) {
    throw new Error("Schema should not be null");
  }

  /**
   * validate object against the schema
   */
  switch (args.schema.type) {
    case "number":
      return validateNumber({ ...args, schema: args.schema as NumberSchema<P> }) as Result<T>;
    case "string":
      return validateString({ ...args, schema: args.schema as StringSchema<P> }) as Result<T>;
    case "boolean":
      return validateBoolean({ ...args, schema: args.schema as BooleanSchema<P> }) as Result<T>;
    case "array":
      return validateArray({ value: args.value as any, parent: args.parent, schema: args.schema as any }) as Result<T>;
    case "object":
      return validateObject({ ...args, schema: args.schema }) as Result<T>;
    case "date":
      return validateDate({ ...args, schema: args.schema as DateSchema<P> }) as Result<T>;
  }
  throw new Error(`Unhandled data type`);
};

// public api
export const validate = <T>(value: T, schema: Schema<T>): Result<T> => {
  return _validate({ value, parent: null as any, schema });
};
