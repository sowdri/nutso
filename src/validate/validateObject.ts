import { ObjectResult } from "../models/result/ObjectResult";
import { ObjectSchema } from "../models/schema/ObjectSchema";
import { optionalFlagValidator } from "../utils/optionalFlagValidator";
import { isNil } from "../utils/typeChecker";
import { _validate } from "./validate";

export const isRegex = (str: string) => {
  return str.startsWith("/") && str.endsWith("/");
};

export const getRegex = (str: string) => {
  return new RegExp(`^${str.slice(1, -1)}$`);
};

export const validateObject = <T extends { [key: string]: any }, R, P>(args: {
  value: T | null;
  root: R;
  parent: P;
  schema: ObjectSchema<T, R, P>;
}): ObjectResult<T> => {
  //
  const { value, schema } = args;

  const result: ObjectResult<T> = {
    isValid: true,
    errorMessage: "",
    properties: {} as any,
    errorPath: [],
  };

  // isnil
  if (isNil(value)) {
    return { ...optionalFlagValidator({ ...args, flag: schema.optional }), properties: {} as any };
  }

  const obj = value as T;
  const processedFields: string[] = [];

  // for each static key, validate
  for (let field in schema.properties) {
    if (isRegex(field)) continue;
    result.properties[field] = _validate({
      ...args,
      value: value ? value[field] : null,
      parent: value,
      schema: schema.properties[field],
    });
    processedFields.push(field);
  }

  // for each regex validate all matching keys
  for (let field in schema.properties) {
    if (!isRegex(field)) continue; // not-regex
    const regex = getRegex(field);
    // for each key
    for (let key in obj) {
      /**
       * Already processed field. This will happen if the regex matches a wide range of fields, like ".*"
       * and the properties also defines specific fields like "red": {}, "blue": {} etc ...
       */
      if (processedFields.includes(key)) continue;
      if (!regex.test(key)) continue;
      result.properties[key] = _validate({
        ...args,
        value: value ? value[key] : null,
        parent: value,
        schema: schema.properties[field],
      });
      processedFields.push(key);
    }
  }

  // if this node is valid, then check if all of it's children are valid
  // because the node is invalid, if any of it's children are invalid
  if (result.isValid) {
    for (let field of processedFields) {
      const property = result.properties[field];
      if (!property.isValid) {
        result.isValid = false;
        result.errorMessage = property.errorMessage;
        result.errorPath = [field, ...property.errorPath];
        break;
      }
    }
  }

  return result;
};
