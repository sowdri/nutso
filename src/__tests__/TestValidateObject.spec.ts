import { ObjectSchema } from "../models/schema/ObjectSchema";
import { Schema } from "../models/schema/Schema";
import { validate } from "../validate/validate";
import { validateObject, isRegex, getRegex } from "../validate/validateObject";

test(`isRegex`, () => {
  const test = isRegex("/.*/");
  expect(test).toBe(true);
});

test(`getRegex`, () => {
  const regex = getRegex("/.*/");
  expect(regex.test("foo")).toBe(true);
});

test(`Basic`, () => {
  type Customer = {
    name: "";
  };

  const obj: Customer = { name: "" };
  const schema: ObjectSchema<Customer> = {
    type: "object",
    properties: {
      name: {
        type: "string",
        minLength: 3,
      },
    },
  };

  const result1 = validateObject({ value: obj, root: obj, schema, parent: undefined as any });
  expect(result1.isValid).toBe(false);

  const result2 = validate(obj, schema);
  expect(result2.isValid).toBe(false);
});

test(`Address is optional in schema and undefined in object`, () => {
  type Customer = {
    name: string;
    address?: {
      city: string;
    };
  };

  const obj: Customer = { name: "John" };
  const schema: Schema<Customer> = {
    type: "object",
    properties: {
      name: {
        type: "string",
        minLength: 3,
      },
      address: {
        type: "object",
        optional: true,
        properties: {
          city: {
            type: "string",
          },
        },
      },
    },
  };

  const result = validate(obj, schema);
  expect(result).toMatchSnapshot();
});
