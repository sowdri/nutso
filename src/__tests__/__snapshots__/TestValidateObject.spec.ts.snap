// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Address is optional in schema and undefined in object 1`] = `
{
  "errorMessage": "",
  "errorPath": [],
  "isValid": true,
  "properties": {
    "address": {
      "errorMessage": "",
      "errorPath": [],
      "isValid": true,
      "properties": {},
    },
    "name": {
      "errorMessage": "",
      "errorPath": [],
      "isValid": true,
    },
  },
}
`;
